internal/provider/bms/bms_types.go

```diff
@@ -0,0 +1,139 @@
+package bms
+
+import (
+	"context"
+
+	timeoutsD "github.com/hashicorp/terraform-plugin-framework-timeouts/datasource/timeouts"
+
+	supertypes "github.com/FrangipaneTeam/terraform-plugin-framework-supertypes"
+
+	v1 "github.com/orange-cloudavenue/cloudavenue-sdk-go/v1"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/pkg/utils"
+)
+
+type (
+	bmsModelDatasource struct {
+		ID       supertypes.StringValue                                   `tfsdk:"id"`
+		Timeouts timeoutsD.Value                                          `tfsdk:"timeouts"`
+		Env      supertypes.SetNestedObjectValueOf[bmsModelDatasourceEnv] `tfsdk:"env"`
+	}
+
+	bmsModelDatasourceEnv struct {
+		Network supertypes.SetNestedObjectValueOf[bmsModelDatasourceNetwork] `tfsdk:"network"`
+		BMS     supertypes.SetNestedObjectValueOf[bmsModelDatasourceBMS]     `tfsdk:"bms"`
+	}
+
+	bmsModelDatasourceNetwork struct {
+		VLANID supertypes.StringValue `tfsdk:"vlan_id"`
+		Subnet supertypes.StringValue `tfsdk:"subnet"`
+		Prefix supertypes.StringValue `tfsdk:"prefix"`
+	}
+
+	bmsModelDatasourceBMS struct {
+		Hostname          supertypes.StringValue                                             `tfsdk:"hostname"`
+		BMSType           supertypes.StringValue                                             `tfsdk:"bms_type"`
+		OS                supertypes.StringValue                                             `tfsdk:"os"`
+		BiosConfiguration supertypes.StringValue                                             `tfsdk:"bios_configuration"`
+		Storage           supertypes.SingleNestedObjectValueOf[bmsModelDatasourceBMSStorage] `tfsdk:"storage"`
+	}
+
+	bmsModelDatasourceBMSStorage struct {
+		Local  supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"local"`
+		System supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"system"`
+		Data   supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"data"`
+		Shared supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"shared"`
+	}
+
+	bmsModelDatasourceBMSStorageGen struct {
+		Size         supertypes.StringValue `tfsdk:"size"`
+		StorageClass supertypes.StringValue `tfsdk:"storage_class"`
+	}
+)
+
+// NewbmsModelDatasource returns a new bmsModelDatasource.
+func NewBMSModelDatasource(ctx context.Context) *bmsModelDatasource {
+	return &bmsModelDatasource{
+		ID:       supertypes.NewStringNull(),
+		Timeouts: timeoutsD.Value{},
+		Env:      supertypes.NewSetNestedObjectValueOfNull[bmsModelDatasourceEnv](ctx),
+	}
+}
+
+// SetID sets the ID.
+func (m *bmsModelDatasource) SetID(id *string) {
+	m.ID.Set(*id)
+}
+
+// SetNetwork sets the Network of BMS listed.
+func (m *bmsModelDatasource) SetNetwork(ctx context.Context, bms *v1.BMS) (net []*bmsModelDatasourceNetwork, err error) {
+	networks := bms.GetNetworks()
+	if len(networks) == 0 {
+		return make([]*bmsModelDatasourceNetwork, 0), nil
+	}
```

Comment:
This is unnessarry. if networks have not entry the for loop is ignored.

Reply:
Fix